experimental/Schemes/CoherentSheaves.jl

```diff
@@ -947,6 +1089,215 @@ function Base.show(io::IO, M::PushforwardSheaf)
+  f::AbsCoveredSchemeMorphism
+  OOX::StructureSheafOfRings # the sheaf of rings in the domain
+  OOY::StructureSheafOfRings # the sheaf of rings in the codomain
+  M::AbsCoherentSheaf        # the sheaf of modules on Y
+  pullback_of_sections::IdDict{AbsSpec, Union{Hecke.Map, Nothing}} # a dictionary caching the natural 
+                                                                   # pullback maps along the maps in the `covering_morphism` of f 
+  F::PreSheafOnScheme        # the internal caching instance doing the bookkeeping
+
+  function PullbackSheaf(f::AbsCoveredSchemeMorphism, M::AbsCoherentSheaf)
+    X = domain(f)
+    Y = codomain(f)
+    Y === scheme(M) || error("sheaf must be defined over the domain of the embedding")
+    OOY = sheaf_of_rings(M)
+    OOX = OO(X)
+    fcov = covering_morphism(f)::CoveringMorphism
+    CX = domain(fcov)::Covering
+    CY = codomain(fcov)::Covering
+    pullbacks = IdDict{AbsSpec, Hecke.Map}()
+
+    ### Production of the modules on open sets.
+    #
+    # Since the morphism f might have an underlying CoveringMorphism ϕ with 
+    # a non-trivial refinement of the `default_covering` of X as a domain, 
+    # we can not expect to easily produce f^*(M) on the `affine_charts` of X. 
+    # Instead, we can produce it on affine opens U ⊂ X which are hanging 
+    # below the patches in `domain(ϕ)`.
+    #
+    # For everything else, we proceed as in case 3) of the general 
+    # SheafOfModules, see above. 
+    #
+    # Again, this case is not implemented for the time being. 
+
+    function production_func(FF::AbsPreSheaf, U::AbsSpec)
+      # See whether U is a patch of the domain covering and pull back directly
+      if haskey(morphisms(fcov), U)
+        floc = morphisms(fcov)[U]
+        MU, map = change_base_ring(pullback(floc), M(codomain(floc)))
+        pullbacks[U] = map
+        return MU
+      end
+
+      # We are in case 3).
+      error("case not implemented")
+    end
+
+    function production_func(FF::AbsPreSheaf, 
+        U::Union{<:PrincipalOpenSubset, <:SimplifiedSpec}
+      )
+      # See whether U is a patch of the domain covering and pull back directly
+      if haskey(morphisms(fcov), U)
+        floc = morphisms(fcov)[U]
+        MU, map = change_base_ring(pullback(floc), M(codomain(floc)))
+        pullbacks[U] = map
+        return MU
+      end
+
+      # If not, check whether we are hanging below such a patch in the 
+      # refinement tree.
+      if some_ancestor(y->any(x->(x===y), patches(domain(fcov))), U)
+        V = __find_chart(U, domain(fcov))
+        MU, res = change_base_ring(OOX(V, U), FF(V))
+        add_incoming_restriction!(FF, V, MU, res)
+        return MU
+      end
+
+      # We are in case 3)
+      error("case not implemented")
+    end
+
+    ### Restriction for pulled back sheaves of modules
+    #
+    # For U ⊂ V ⊂ X, f : X → Y, M on Y and F = f^*M we do the following. 
+    # Let ϕ be the `covering_morphism` behind f. 
```

Comment:
Do not just fix notation here, but write out clearly in mathematical terms, what the goal is, before going into the case below.